File: src/main/java/me/testaccount666/migration/plugins/essentials/BalanceMigrator.java
package me.testaccount666.migration.plugins.essentials;

import me.testaccount666.serversystem.ServerSystem;
import net.ess3.api.MaxMoneyException;
import net.ess3.api.events.UserBalanceUpdateEvent;

import java.util.logging.Level;

public class BalanceMigrator extends AbstractMigrator {

    @Override
    public int migrateFrom() {
        var count = 0;

        var userManager = userManager();
        var essentials = essentials();
        var uuids = essentials.getUsers().getAllUserUUIDs();

        for (var uuid : uuids) {
            var userOptional = userManager.getUser(uuid);
            if (userOptional.isEmpty()) {
                ServerSystem.getLog().warning("Couldn't find user '${uuid}', skipping balance migration!");
                continue;
            }

            var essentialsUser = essentials.getUser(uuid);
            var user = userOptional.get().getOfflineUser();
            var bankAccount = user.getBankAccount();

            bankAccount.setBalance(essentialsUser.getMoney());

            user.save();
            count += 1;
        }

        return count;
    }

    @Override
    public int migrateTo() {
        var count = 0;

        var userManager = userManager();

        for (var player : offlinePlayers()) {
            var uuid = player.getUniqueId();

            var userOptional = userManager.getUser(uuid);
            if (userOptional.isEmpty()) {
                ServerSystem.getLog().warning("Couldn't find user '${uuid}', skipping balance migration!");
                continue;
            }

            var user = userOptional.get().getOfflineUser();
            var bankAccount = user.getBankAccount();

            var essentials = essentials();

            ensureUserDataExists(uuid);
            var essentialsUser = essentials.getUser(uuid);

            try {
                essentialsUser.setMoney(bankAccount.getBalance(), UserBalanceUpdateEvent.Cause.SPECIAL);
                count += 1;
            } catch (MaxMoneyException exception) {
                var userName = user.getName().orElse("Unknown");

                ServerSystem.getLog().log(Level.WARNING, "Couldn't migrate balance for '${uuid}' (${userName}) with balance '${bankAccount.getBalance()}'", exception);
            }
        }

        return count;
    }
}

File: src/main/kotlin/me/testaccount666/migration/plugins/essentials/BalanceMigrator.kt
package me.testaccount666.migration.plugins.essentials

import me.testaccount666.serversystem.ServerSystem.Companion.log
import net.ess3.api.events.UserBalanceUpdateEvent
import java.util.logging.Level

class BalanceMigrator : AbstractMigrator() {
    override fun migrateFrom(): Int {
        val count = essentials.users.allUserUUIDs.count { uuid ->
            runCatching {
                val cachedUser = userManager.getUserOrNull(uuid) ?: run {
                    log.warning("Couldn't find user '${uuid}', skipping balance migration!")
                    return@count false
                }

                val essentialsUser = essentials.getUser(uuid)
                val user = cachedUser.offlineUser
                val bankAccount = user.bankAccount

                bankAccount.balance = essentialsUser.money

                user.save()
                return@runCatching true
            }.onFailure { log.log(Level.WARNING, "Couldn't migrate balance for '${uuid}'", it) }.getOrDefault(false)
        }

        return count
    }

    override fun migrateTo(): Int {
        val count = offlinePlayers().count { player ->
            val uuid = player.uniqueId
            runCatching {
                val cachedUser = userManager.getUserOrNull(uuid) ?: run {
                    log.warning("Couldn't find user '${uuid}', skipping balance migration!")
                    return@count false
                }

                val user = cachedUser.offlineUser
                val bankAccount = user.bankAccount

                ensureUserDataExists(uuid)
                val essentialsUser = essentials.getUser(uuid)

                essentialsUser.setMoney(bankAccount.balance, UserBalanceUpdateEvent.Cause.SPECIAL)
                return@runCatching true
            }.onFailure { log.log(Level.WARNING, "Couldn't migrate balance for '${uuid}'", it) }.getOrDefault(false)
        }

        return count
    }
}
